File: AsyncEnumerable-HttpServer/csharp/HttpServer.cs
using System.Collections.Generic;
using System.Net;
using System.Runtime.CompilerServices;
using System.Threading;

namespace Bomret.Dotnet.Examples.AsyncEnumerableHttpServer.CSharp
{
  public static class HttpServer
  {
    public struct Options
    {
      public string Hostname;
      public ushort Port;
    }

    public static async IAsyncEnumerable<HttpListenerContext> Serve(Options options, [EnumeratorCancellation] CancellationToken cancellationToken)
    {
      using var listener = new HttpListener();

      string url = $"http://{options.Hostname}:{options.Port}/";

      listener.Prefixes.Add(url);
      listener.Start();

      while (!cancellationToken.IsCancellationRequested)
      {
        var context = await listener.GetContextAsync().ConfigureAwait(false);

        yield return context;
      }
    }
  }
}

File: AsyncEnumerable-HttpServer/fsharp/HttpServer.fs
namespace Bomret.Dotnet.Examples.AsyncEnumerableHttpServer.FSharp

open System.Net
open System.Threading

open FSharp.Control

module HttpServer =
    [<Struct>]
    type Options = { Hostname: string; Port: uint16 }

    let Serve (options: Options, cancellationToken: CancellationToken): IAsyncEnumerable<HttpListenerContext> =
        asyncSeq {
            use listener = new HttpListener()

            let url =
                sprintf "http://%s:%i/" options.Hostname options.Port

            listener.Prefixes.Add(url)
            listener.Start()

            while not cancellationToken.IsCancellationRequested do
                let! context = listener.GetContextAsync() |> Async.AwaitTask

                yield context
        }
